core/block.js

```diff
@@ -741,21 +741,31 @@ Blockly.Block.prototype.getChildren = function(ordered) {
  * @param {Blockly.Block} newParent New parent block.
  */
 Blockly.Block.prototype.setParent = function(newParent) {
-  if (newParent == this.parentBlock_) {
+  if (newParent === this.parentBlock_) {
     return;
   }
+
+  // Check that block is connected to new parent if new parent is not null and
+  //    that block is not connected to superior one if new parent is null.
+  var connection = this.previousConnection || this.outputConnection;
+  var isConnected = !!(connection && connection.targetBlock());
+
+  if (isConnected && newParent && connection.targetBlock() !== newParent) {
+    throw Error('Block connected to superior one that is not new parent.');
+  } else if (!isConnected && newParent) {
+    throw Error('Block not connected to new parent.');
+  } else if (isConnected && !newParent) {
+    throw Error('Cannot set parent to null while block is still connected to' +
+        ' superior block.');
+  }
+
   if (this.parentBlock_) {
```

Comment:
I know this isn't exactly related to what you're doing, but do you think this if-statement and the below one could be combined? It seems like they're doing the same things (updating child blocks and top blocks).

Reply:
I could be missing something, but I'm not sure how to cleanly group those two if statements given the overlap of the following conditional tasks.

```
                             Remove from old parent  Add to new parent Remove top block  Add top block
Has old/new parent block              Y                     Y                 N               N
Only has old parent block             Y                     N                 N               Y
Only has new parent block             N                     Y                 Y               N               
```

So for an example, having an if statement selecting only for the first case would mean duplicating code to remove/add parents.  But maybe that's fine?  I could also remove the second else and instead conditionally add a top block if there's no new parent by nesting an if statement inside the first, but I'm not sure that makes the code cleaner?  Thoughts?

I did however remove the unnecessary statement setting `this.parentBlock_` to `null` (and associated comment) as it unconditionally gets set to `newParent` later.